llama_index/callbacks/base.py

```diff
@@ -161,9 +194,43 @@ def end_trace(
 
     def _reset_trace_events(self) -> None:
         """Helper function to reset the current trace."""
+
         self._trace_map = defaultdict(list)
-        self._trace_event_stack = [BASE_TRACE_EVENT]
+        self._trace_event_stack.set([BASE_TRACE_EVENT])
 
     @property
     def trace_map(self) -> Dict[str, List[str]]:
         return self._trace_map
+
+
+class EventContext:
+    """
+    Simple wrapper to call callbacks on event starts and ends
+    with an event type and id.
+    """
+
+    def __init__(
+        self,
+        callback_manager: CallbackManager,
+        event_type: CBEventType,
+        event_id: Optional[str] = None,
+    ):
+        self._callback_manager = callback_manager
+        self._event_type = event_type
+        self._event_id = event_id or str(uuid.uuid4())
+        self.started = False
+        self.finished = False
+
+    def on_start(self, payload: Optional[Dict[str, Any]] = None, **kwargs: Any) -> None:
+        if not self.started:

```

Comment:
very nit: I wonder if we should log a warning instead of ignoring it slightly when an event has already started, but caller called `on_start` again.

Reply:
Yea logging a warning is good